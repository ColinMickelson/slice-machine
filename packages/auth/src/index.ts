export { createPrismicAuth } from "./createPrismicAuth";

export { createPrismicAuthMiddleware } from "./createPrismicAuthMiddleware";

export type { PrismicAuth } from "./PrismicAuth";

export type { PrismicUserProfile } from "./types";

export {
	UnauthenticatedError,
	isUnauthenticatedError,
	UnexpectedDataError,
	isUnexpectedDataError,
} from "./errors";
