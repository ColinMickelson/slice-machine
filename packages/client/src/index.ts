import axios, { AxiosError, AxiosPromise, AxiosRequestHeaders } from "axios";
import * as t from "io-ts";

import { UserProfile } from "@slicemachine/core/build/models";
import { SharedSlice } from "@prismicio/types-internal/lib/customtypes/widgets/slices";
import { CustomType } from "@prismicio/types-internal/lib/customtypes/CustomType";

import type { ClientError, UploadParameters } from "./models";
import {
  ApplicationMode,
  getStatus,
  getMessage,
  ApisEndpoints,
  ProductionApisEndpoints,
  StageApisEndpoints,
  AclCreateResult,
} from "./models";

import { getAndValidateResponse } from "./utils";
import { upload } from "./utils/upload";

// exporting models to be used with the Client.
export * from "./models";

// exporting utils so the client calls can be replicated for speficic calls to a consumer.
export * from "./utils";

// exporting the client itself.
export class Client {
  apisEndpoints: ApisEndpoints;
  repository: string | null;
  authenticationToken: string;

  constructor(
    applicationMode: ApplicationMode,
    repository: string | null,
    authenticationToken: string
  ) {
    this.repository = repository;
    this.authenticationToken = authenticationToken;

    if (applicationMode === ApplicationMode.PROD)
      this.apisEndpoints = ProductionApisEndpoints;
    else if (applicationMode === ApplicationMode.STAGE)
      this.apisEndpoints = StageApisEndpoints;
    else {
      // Dev
      this.apisEndpoints = ProductionApisEndpoints;
    }
  }

  // setters to provide flexibility
  updateAuthenticationToken(newToken: string): Client {
    this.authenticationToken = newToken;
    return this
  }

  updateRepository(repository: string | null): Client {
    this.repository = repository;
    return this
  }

  // private methods
  _fetch(
    method: "get" | "post",
    url: string,
    data?: Record<string, unknown>,
    headers?: AxiosRequestHeaders
  ): AxiosPromise {
    return axios({
      method,
      url,
      data,
      headers: {
        Authorization: `Bearer ${this.authenticationToken}`,
        "User-Agent": "slice-machine",
        ...(this.repository ? { repository: this.repository } : {}),
        ...headers,
      },
    });
  }

  _get(url: string): AxiosPromise {
    return this._fetch("get", url);
  }

  _post(url: string, data: Record<string, unknown>): AxiosPromise {
    return this._fetch("post", url, data).catch((error: Error | AxiosError) => {
      const status: number = getStatus(error);
      const message: string = getMessage(error);

      // Making sure the error is typed
      const clientError: ClientError = { status, message };
      return Promise.reject(clientError);
    });
  }

  async validateAuthenticationToken(): Promise<boolean> {
    return this._get(
      `${this.apisEndpoints.Authentication}validate?token=${this.authenticationToken}`
    )
      .then(() => true)
      .catch((error: Error | AxiosError) => {
        const status: number = getStatus(error);
        if (status == 401) return false;

        const message: string = getMessage(error);

        // Making sure the error is typed
        const clientError: ClientError = { status, message };
        return Promise.reject(clientError);
      });
  }

  async refreshAuthenticationToken(): Promise<string> {
    return getAndValidateResponse<string>(
      this._get(
        `${this.apisEndpoints.Authentication}refreshtoken?token=${this.authenticationToken}`
      ),
      "refreshed authentication token",
      t.string
    ).then((newToken) => {
      this.updateAuthenticationToken(newToken);
      return newToken;
    });
  }

  async profile(): Promise<UserProfile> {
    return getAndValidateResponse<UserProfile>(
      this._get(`${this.apisEndpoints.Users}profile`),
      "user profile",
      UserProfile
    );
  }

  async getSlices(): Promise<SharedSlice[]> {
    return getAndValidateResponse<SharedSlice[]>(
      this._get(`${this.apisEndpoints.Models}slices`),
      "shared slices",
      t.array(SharedSlice)
    );
  }

  async getCustomTypes(): Promise<CustomType[]> {
    return getAndValidateResponse<CustomType[]>(
      this._get(`${this.apisEndpoints.Models}customtypes`),
      "custom types",
      t.array(CustomType)
    );
  }

  async insertCustomType(customType: CustomType): Promise<AxiosPromise> {
    return this._post(
      `${this.apisEndpoints.Models}customtypes/insert`,
      customType
    );
  }

  async updateCustomType(customType: CustomType): Promise<AxiosPromise> {
    return this._post(
      `${this.apisEndpoints.Models}customtypes/update`,
      customType
    );
  }

  async insertSlice(sharedSlice: SharedSlice): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.Models}slices/insert`, sharedSlice);
  }

  async updateSlice(sharedSlice: SharedSlice): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.Models}slices/update`, sharedSlice);
  }

  async createImagesAcl(): Promise<AclCreateResult> {
    return getAndValidateResponse<AclCreateResult>(
      this._get(`${this.apisEndpoints.AclProvider}create`),
      "acl",
      AclCreateResult
    );
  }

  async deleteImagesFolderAcl(sliceName: string): Promise<AxiosPromise> {
    return this._post(`${this.apisEndpoints.AclProvider}delete-folder`, {
      sliceName,
    });
  }

  async uploadImageAcl(params: UploadParameters): Promise<number | undefined> {
    return upload(params);
  }
}
