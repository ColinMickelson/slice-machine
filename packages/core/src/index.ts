export * as FileSystem from "./filesystem";
export * as Utils from "./utils";
export * from "./core";
export * as Models from "./models";
export * as Libraries from "./libraries";
export * as Mocks from "./libraries/mocks";
export * from "./auth";
