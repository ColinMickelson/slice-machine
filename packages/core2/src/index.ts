export {
	SliceMachineManager,
	createSliceMachineManager,
} from "./createSliceMachineManager";

export {
	CreateSliceMachineManagerMiddlewareArgs,
	createSliceMachineManagerMiddleware,
} from "./createSliceMachineManagerMiddleware";

export {
	PrismicAuthManager,
	PrismicUserProfile,
	createPrismicAuthManager,
} from "./createPrismicAuthManager";

export {
	CreatePrismicAuthManagerMiddlewareArgs,
	PrismicAuthCheckStatusResponse,
	createPrismicAuthManagerMiddleware,
} from "./createPrismicAuthManagerMiddleware";
