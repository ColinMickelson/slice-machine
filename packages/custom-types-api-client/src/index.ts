export { createCustomTypesAPIClient } from "./createCustomTypesAPIClient";

export type { CustomTypesAPIClient } from "./CustomTypesAPIClient";

export {
	InvalidPayloadError,
	ForbiddenError,
	ConflictError,
	NotFoundError,
	PrismicError,
} from "@prismicio/custom-types-client";
