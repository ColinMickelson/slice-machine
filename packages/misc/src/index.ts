// TODO: Figure out where libs should be
export { decode } from "./decode";
export type { DecodeReturnType } from "./decode";
export { DecodeError } from "./DecodeError";
export {
	readPrismicrc,
	writePrismicrc,
	updatePrismicrc,
	Prismicrc,
} from "./prismicrc";
export {
	getAllStableSliceMachineVersions,
	getReleaseNotesForVersion,
} from "./versions";

// Errors
export { SliceMachineError, isSliceMachineError } from "./SliceMachineError";
export { InternalError, isInternalError } from "./InternalError";

// Constants
export { ApplicationMode } from "./ApplicationMode";
export { APIEndpoints } from "./APIEndpoints";
export { APITokens } from "./APITokens";
