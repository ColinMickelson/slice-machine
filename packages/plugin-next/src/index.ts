import type { Plugin } from "@slicemachine/plugin-middleware";
export { slice, story, index } from "@slicemachine/plugin-react";

export const framework: Plugin["framework"] = "next";
