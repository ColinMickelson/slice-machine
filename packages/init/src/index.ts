#!/usr/bin/env node

import { Utils, FileSystem } from "@slicemachine/core";
import { Tracker } from "./utils/tracker";
import {
  installRequiredDependencies,
  validatePkg,
  maybeExistingRepo,
  createRepository,
  loginOrBypass,
  configureProject,
  displayFinalMessage,
  detectFramework,
  installLib,
} from "./steps";
import { findArgument } from "./utils";

async function init() {
  const cwd = findArgument(process.argv, "cwd") || process.cwd();
  const base = findArgument(process.argv, "base") || Utils.CONSTS.DEFAULT_BASE;
  const lib: string | undefined = findArgument(process.argv, "library");
  const branch: string | undefined = findArgument(process.argv, "branch");
  const tracking =
    findArgument(process.argv, "tracking") === "false" ? false : true;

  console.log(
    Utils.purple(
      "You're about to configure Slicemachine... Press ctrl + C to cancel"
    )
  );

  // verify package.json file exist
  validatePkg(cwd);

  // login
  const user = await loginOrBypass(base);
  if (!user) throw new Error("The user should be logged in!");

  // retrieve tokens for api calls
  const config = FileSystem.PrismicSharedConfigManager.get();

  // detect the framework used by the project
  const frameworkResult = await detectFramework(cwd);

  // select the repository used with the project.
  const { existing, name } = await maybeExistingRepo(
    config.cookies,
    cwd,
    config.base
  );

  if (!existing) {
    await createRepository(name, frameworkResult.value, config);
  }

  const tracker = Tracker.build(
    "JfTfmHaATChc4xueS7RcCBsixI71dJIJ",
    name,
    {
      userId: user.userId,
    },
    tracking
  );

  // install the required dependencies in the project.
  await installRequiredDependencies(cwd, frameworkResult.value);

  const sliceLibPath = lib
    ? await installLib(tracker, cwd, lib, branch)
    : undefined;

  // configure the SM.json file and the json package file of the project..
  configureProject(cwd, base, name, frameworkResult, sliceLibPath, tracking);

  // ask the user to run slice-machine.
  displayFinalMessage(cwd);
}

try {
  void init();
} catch (error) {
  if (error instanceof Error) Utils.writeError(error.message);
  else console.error(error);
}
