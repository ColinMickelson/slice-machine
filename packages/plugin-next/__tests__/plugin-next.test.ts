import { slice } from "../src";

describe("@slicemachine/plugin-next", () => {
  test("#slice", () => {
    expect(slice()).toBeUndefined();
  });
});
