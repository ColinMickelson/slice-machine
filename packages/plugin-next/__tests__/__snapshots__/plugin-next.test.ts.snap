// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@slicemachine/plugin-next #slice 1`] = `
Object {
  "data": "import React from 'react'
import { RichText } from 'prismic-reactjs'
  
const Foo = ({ slice }) => (
  <section>
    <span className=\\"title\\">
      { slice.primary.title ? (
        <RichText render={slice.primary.title}/
      ) : (
        <h2>Template slice, update me!</h2>
      )}
    </span>
      { slice.primary.description ? (
        <RichText render={slice.primary.description}/>
      ) : (
         <p>start by editing this slice from inside Slice Machine!</p>
      )}
      <style jsx>{\`
          section {
            max-width: 600px;
            margin: 4em auto;
            text-align: center;
          }
          .title {
            color: #8592e0;
          }
      \`}</style>
    </section>
  )
  
  export default Foo
",
  "filename": "index.js",
}
`;

exports[`@slicemachine/plugin-next story 1`] = `
Object {
  "data": "import MyComponent from '../slices/foo';
export default { title: 'FooBar' }

export const _DefaultSlice = () => <MyComponent slice={{\\"id\\":\\"default-slice\\",\\"name\\":\\"Default slice\\",\\"docURL\\":\\"...\\",\\"version\\":\\"sktwi1xtmkfgx8626\\",\\"description\\":\\"FooSlice\\",\\"primary\\":{\\"title\\":{\\"type\\":\\"StructuredText\\",\\"config\\":{\\"single\\":\\"heading1\\",\\"label\\":\\"Title\\",\\"placeholder\\":\\"This is where it all begins...\\"}},\\"description\\":{\\"type\\":\\"StructuredText\\",\\"config\\":{\\"single\\":\\"paragraph\\",\\"label\\":\\"Description\\",\\"placeholder\\":\\"A nice description of your feature\\"}},\\"\\":{\\"config\\":{\\"label\\":\\"\\",\\"constraint\\":{},\\"thumbnails\\":[]},\\"type\\":\\"Image\\"}},\\"items\\":{}}} />
_DefaultSlice.storyName = 'Default slice'

",
  "filename": "index.stories.js",
}
`;
